homeassistant/components/upnp/config_flow.py

```diff
@@ -0,0 +1,160 @@
+
+    def _store_discovery_info(self, discovery_info):
+        """Add discovery info."""
+        udn = discovery_info['udn']
+        self.hass.data[DOMAIN]['discovered'][udn] = discovery_info
+
+    def _auto_config_settings(self):
+        """Check if auto_config has been enabled."""
+        return self.hass.data[DOMAIN]['auto_config']
+
+    async def async_step_discovery(self, discovery_info):
+        """
+        Handle a discovered UPnP/IGD.
+
+        This flow is triggered by the discovery component. It will check if the
+        host is already configured and delegate to the import step if not.
+        """
+        ensure_domain_data(self.hass)
+
+        # store discovered device
+        discovery_info['friendly_name'] = \
+            '{} ({})'.format(discovery_info['host'], discovery_info['name'])
+        self._store_discovery_info(discovery_info)
+
+        # ensure not already discovered/configured
+        udn = discovery_info['udn']
+        if udn in self._configured_upnp_igds:
+            return self.async_abort(reason='already_configured')
+
+        # auto config?
+        auto_config = self._auto_config_settings()
+        if auto_config['active']:
+            import_info = {
+                'name': discovery_info['friendly_name'],
+                'enable_sensors': auto_config['enable_sensors'],
+                'enable_port_mapping': auto_config['enable_port_mapping'],
+            }
+
+            return await self._async_save_entry(import_info)
+
+        return await self.async_step_user()
+
+    async def async_step_user(self, user_input=None):
+        """Manual set up."""
+        ensure_domain_data(self.hass)
+
+        # if user input given, handle it
+        user_input = user_input or {}
+        if 'name' in user_input:
+            if not user_input['enable_sensors'] and \
+               not user_input['enable_port_mapping']:
+                return self.async_abort(reason='no_sensors_or_port_mapping')
+
+            # ensure not already configured
+            configured_names = [
+                entry['friendly_name']
+                for udn, entry in self._discovered_upnp_igds.items()
+                if udn in self._configured_upnp_igds
+            ]
+            if user_input['name'] in configured_names:
+                return self.async_abort(reason='already_configured')
+
+            return await self._async_save_entry(user_input)
+
+        # let user choose from all discovered, non-configured, UPnP/IGDs
+        names = [
+            entry['friendly_name']
+            for udn, entry in self._discovered_upnp_igds.items()
+            if udn not in self._configured_upnp_igds
+        ]
+        if not names:
+            return self.async_abort(reason='no_devices_discovered')
+
+        return self.async_show_form(
+            step_id='user',
+            data_schema=vol.Schema(
+                OrderedDict([
+                    (vol.Required('name'), vol.In(names)),
```

Comment:
Any reason we would have users pick and not just load them all?

Reply:
I can image, if you have multiple internet connections (Small Offices/Home Offices), you don't want all connections to create port mappings to hass.

Reply:
In #17229 there is already that use case (equipment is an Access Point)

Reply:
okay so now I am more confused haha . I thought this config entry was about sensors but it's about port mapping? In that case it makes sense.

Reply:
It is both. You can use the traffic-sensors and/or create a port-mapping.